components/salsa-macro-rules/src/setup_input_struct.rs

```diff
@@ -204,6 +229,72 @@ macro_rules! setup_input_struct {
                     })
                 }
             }
+
+            impl $zalsa_struct::HasBuilder for $Struct {
+                type Builder = builder::$Builder;
+            }
+
+            mod builder {
+                use super::*;
+
+                use salsa::plumbing as $zalsa;
+                use $zalsa::input as $zalsa_struct;
+
+                // These are standalone functions instead of methods on `Builder` to prevent
+                // that the enclosing module can call them.
+                pub(super) fn new_builder($($field_id: $field_ty),*) -> $Builder {
+                    $Builder {
+                        fields: ($($field_id,)*),
+                        durabilities: [$zalsa::Durability::default(); $N],
+                    }
+                }
+
+                pub(super) fn builder_into_inner(builder: $Builder, revision: $zalsa::Revision) -> (($($field_ty,)*), $zalsa::Array<$zalsa::Stamp, $N>) {
+                    let stamps = $zalsa::Array::new([
+                        $($zalsa::stamp(revision, builder.durabilities[$field_index])),*
+                    ]);
+
+                    (builder.fields, stamps)
+                }
+
+                pub struct $Builder {
+                    /// The field values.
+                    fields: ($($field_ty,)*),
+
+                    /// The durabilities per field.
+                    durabilities: [$zalsa::Durability; $N],
```

Comment:
I'm a bit confused. For some reason, this compiles in the test code:

```rust
    let input = MyInput::builder(true).durabilities;
```

I would have expected this to fail because the `Builder` is defined in its own module and the field isn't `pub(super)`.